### PR TITLE
[MINDEXER-143] Implement Search API and backends for it.

### PR DESCRIPTION
**Pull request by cstamas**
High level changes:
* new package for Search API: `org.apache.maven.search` (ideally i'd put this into new git repo: maven-search)
* implement search-api
* implement search-backend-indexer and search-backend-smo
* currently they behave "nearly identical" but there are most probably more cases for tuning

---

https://issues.apache.org/jira/browse/MINDEXER-143

### CODE REVIEW

File: search-backend-indexer/src/main/java/org/apache/maven/index/search/backend/indexer/internal/IndexerCoreSearchBackendImpl.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    {
+        return indexingContext;
+    }
+
+    @NotNull
+    @Override
+    public IndexerCoreSearchResponse search( SearchRequest searchRequest ) throws IOException
+    {
+        // if GA present in query: doing flat, otherwise grouped search to mimic SMO
+        HashSet<Field> searchedFields = new HashSet<>();
```
Comment (michael-o) [inline search-backend-indexer/src/main/java/org/apache/maven/index/search/backend/indexer/internal/IndexerCoreSearchBackendImpl.java:105]: why is this typed and the upcoming collections not?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: search-backend-smo/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Indexer Search SMO Backend
+==========================
+
+Search API SMO (https://search.maven.org/) backend implementation.
+
+By default uses GSON only, so is Java8, Android and GraalVM (untested) friendly.
```
[the remark this answers is not in the document]
Comment (cstamas) [reply search-backend-smo/README.md:6]: This is wrong, as IMHO 
* GSON is quite lively project
* best fitted to this use case (no OM needed, just picking string/number fields from response JSON) -- ideally would use `org.json` but license disallows it

File: search-backend-indexer/src/main/java/org/apache/maven/index/search/backend/indexer/internal/IndexerCoreSearchBackendImpl.java (hunk 1 above)
Comment (cstamas) [reply search-backend-indexer/src/main/java/org/apache/maven/index/search/backend/indexer/internal/IndexerCoreSearchBackendImpl.java:101]: WDYM?

File: search-api/src/main/java/org/apache/maven/index/search/SEARCH.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * under the License.
+ */
+
+import org.apache.maven.index.search.request.Field.NumberField;
+import org.apache.maven.index.search.request.Field.StringField;
+
+/**
+ * The ontology of Search API, "meta" fields.
+ */
+public final class SEARCH
```
[the remark this answers is not in the document]
Comment (cstamas) [reply search-api/src/main/java/org/apache/maven/index/search/SEARCH.java:28]: As this is a special class (ontology) and only carries "naming" nothing else. Follows same principle that is already present for Maven Indexer ontologies.

File: search-backend-smo/src/main/java/org/apache/maven/index/search/backend/smo/internal/SmoSearchBackendImpl.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import org.apache.maven.index.search.request.Field;
+import org.apache.maven.index.search.request.FieldQuery;
+import org.apache.maven.index.search.request.Paging;
+import org.apache.maven.index.search.request.Query;
+import org.apache.maven.index.search.support.SearchBackendSupport;
+import org.jetbrains.annotations.NotNull;
+
+import static java.util.Objects.requireNonNull;
+
+public class SmoSearchBackendImpl extends SearchBackendSupport implements SmoSearchBackend
```
[the remark this answers is not in the document]
Comment (cstamas) [reply search-backend-smo/src/main/java/org/apache/maven/index/search/backend/smo/internal/SmoSearchBackendImpl.java:52]: Am not encoding any URL with it, but am encoding query parameter values... Still there is `+` vs `%20` issue that will fix

File: search-backend-smo/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (DarthHater) [reply search-backend-smo/README.md:6]: I dunno I kinda think you should use Moshi (gson author, seems pretty awesome) https://proandroiddev.com/goodbye-gson-hello-moshi-4e591116231e

PS Hi @cstamas !

[the remark this answers is not in the document]
Comment (cstamas) [reply search-backend-smo/README.md:6]: Ironically, my initial implementation did use Moshi, but I disliked the fact it had quite a lot of non-trivial dependencies (okio, kotlin libs). gson is simpler IMHO.

**Comment by jira-importer** [comment]
Resolve #631

**Comment by jira-importer** [comment]
Resolve #631

**Pull request closed without merging**